src/v/compression/async_stream_zstd.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "bytes/iobuf.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/aligned_buffer.hh>
+
+#include <memory>
+#include <zstd.h>
+
+namespace compression {
+class async_stream_zstd {
+public:
+    ss::future<iobuf> compress(const iobuf& b) { return do_compress(b); }
+    ss::future<iobuf> uncompress(const iobuf& b) { return do_uncompress(b); }
+    ss::future<iobuf> compress(iobuf&& b) { return do_compress(b); }
+    ss::future<iobuf> uncompress(iobuf&& b) { return do_uncompress(b); }
+
+    static void init_workspace(size_t);
```

Comment:
I think some comments are needed here, regarding the relationship between init and the other other methods, since init is static but init is not.

OTOH, this is probably a good place to ask if init _should_ be static. What it if where just a member function or this method was rolled into the constructor?

Then there would be a clear relationship between the resources allocated by init and those used by the compress/uncompress calls.

Then if you want a "global singleton", in order to share the workspace resources, you can just have a global (thread-local, really) async_stream_ztd object somewhere.

I find this way of creating singletons (write a non-singleton object, then create 1 instance of it) better than having static methods & shared state between object instances, in particular when it comes to testing.

Reply:
Switched to to your recommended non-singleton object then one instance of it pattern. It made a lot more sense here.